rio/transmat/impl/git2/git_transmat.go

```diff
@@ -0,0 +1,205 @@
+				"remote", uri,
+				"reason", pong,
+			)
+		}
+	}
+	if warehouse == nil {
+		panic(&def.ErrWarehouseUnavailable{
+			Msg:    "No warehouses responded!",
+			During: "fetch",
+		})
+	}
+	return warehouse
+}
+
+/*
+	Git transmats plonk down the contents of one commit (or tree) as a filesystem.
+
+	A fileset materialized by git does *not* include the `.git` dir by default,
+	since those files are not themselves part of what's described by the hash.
+
+	Git effectively "filters" out several attributes -- permissions are only loosely
+	respected (execution only), file timestamps are undefined, uid/gid bits
+	are not tracked, xattrs are not tracked, etc.  If you desired defined values,
+	*you must still configure materialization to use a filter* (particularly for
+	file timestamps, since they will otherwise be allowed to vary from one
+	materialization to the next(!)).
+
+	Git also allows for several other potential pitfalls with lossless data
+	transmission: git cannot transmit empty directories.  This can be a major pain.
+	Typical workarounds include creating a ".gitkeep" file in the empty directory.
+	Gitignore files may also inadventantly cause trouble.  Transmat.Materialize
+	will act *consistently*, but it does not overcome these issues in git
+	(doing so would require additional metadata or protocol extensions).
+
+	This transmat is *not* currently well optimized, and should generally be assumed
+	to be re-cloning on all materializations -- specifically, it is not smart
+	enough to recognize requests for different commits and trees from the
+	same repos in order to save reclones.
+*/
+func (t *GitTransmat) Materialize(
+	kind rio.TransmatKind,
+	dataHash rio.CommitID,
+	siloURIs []rio.SiloURI,
+	log log15.Logger,
+	options ...rio.MaterializerConfigurer,
+) rio.Arena {
+	var arena gitArena
+	meep.Try(func() {
+		// Basic validation and config
+		mixins.MustBeType(Kind, kind)
+
+		// Short circut out if we have the whole hash cached.
+		finalPath := t.workArea.getFullCheckoutFinalPath(string(dataHash))
+		if _, err := os.Stat(finalPath); err == nil {
+			arena.workDirPath = finalPath
+			arena.hash = dataHash
+			return
+		}
+
+		warehouse := selectWarehouse(log, siloURIs)
+
+		arena.workDirPath = t.workArea.makeFullCheckoutTempPath(string(dataHash))
+		defer os.RemoveAll(arena.workDirPath)
+
+		log.Info("git: clone starting",
+			"remote", warehouse.url,
+		)
+
+		{
+			started := time.Now()
+			gitClone(log, warehouse.url, CommitId2Hash(dataHash), t.workArea.gitStorageDirs, arena.workDirPath, SubmoduleRecursionDepth)
+			log.Info("git: clone complete",
+				"remote", warehouse.url,
+				"elapsed", time.Since(started).Seconds(),
+			)
+		}
+
+		// Since git doesn't convey permission bits, the default value
+		// should be 1000 (consistent with being accessible under the "routine" policy).
+		// Chown/chmod everything as such.
+		if err := fs.Chownr(arena.workDirPath, git_uid, git_gid); err != nil {
```

Comment:
We should be able to get rid of this post-facto loop if we're now able to use our own `PlaceFile` function consistently! :D